crates/rpc/src/method/subscribe_new_heads.rs

```diff
@@ -0,0 +1,159 @@
+use std::sync::Arc;
+
+use axum::async_trait;
+use pathfinder_common::{BlockId, BlockNumber};
+use tokio::sync::mpsc;
+
+use crate::context::RpcContext;
+use crate::jsonrpc::{RpcError, RpcSubscriptionFlow};
+
+pub struct SubscribeNewHeads;
+
+#[derive(Debug)]
+pub struct Request {
+    block: BlockId,
+}
+
+impl crate::dto::DeserializeForVersion for Request {
+    fn deserialize(value: crate::dto::Value) -> Result<Self, serde_json::Error> {
+        value.deserialize_map(|value| {
+            Ok(Self {
+                block: value.deserialize_serde("block")?,
+            })
+        })
+    }
+}
+
+#[derive(Debug)]
+pub struct Message(Arc<pathfinder_common::BlockHeader>);
+
+impl crate::dto::serialize::SerializeForVersion for Message {
+    fn serialize(
+        &self,
+        serializer: crate::dto::serialize::Serializer,
+    ) -> Result<crate::dto::serialize::Ok, crate::dto::serialize::Error> {
+        crate::dto::BlockHeader(&self.0).serialize(serializer)
+    }
+}
+
+#[async_trait]
+impl RpcSubscriptionFlow for SubscribeNewHeads {
+    type Request = Request;
+    type Notification = Message;
+
+    fn subscription_name() -> &'static str {
+        "starknet_subscriptionNewHeads"
+    }
+
+    fn starting_block(req: &Self::Request) -> BlockId {
+        req.block
+    }
+
+    async fn catch_up(
+        state: &RpcContext,
+        _req: &Self::Request,
+        from: BlockNumber,
+        to: BlockNumber,
+    ) -> Result<Vec<(Self::Notification, BlockNumber)>, RpcError> {
+        let storage = state.storage.clone();
+        let headers = tokio::task::spawn_blocking(move || -> Result<_, RpcError> {
+            let mut conn = storage.connection().map_err(RpcError::InternalError)?;
+            let db = conn.transaction().map_err(RpcError::InternalError)?;
+            db.block_range(from, to).map_err(RpcError::InternalError)
+        })
+        .await
+        .map_err(|e| RpcError::InternalError(e.into()))??;
+        Ok(headers
+            .into_iter()
+            .map(|header| {
+                let block_number = header.number;
+                (Message(header.into()), block_number)
+            })
+            .collect())
+    }
+
+    async fn subscribe(state: RpcContext, tx: mpsc::Sender<(Self::Notification, BlockNumber)>) {
+        let mut rx = state.notifications.block_headers.subscribe();
+        while let Ok(header) = rx.recv().await {
```

Comment:
We should log (on debug?) if the receiver was lagging so that this edge case can be diagnosed by the node operator.